dll/shellext/cabview/extract.cpp

```diff
@@ -0,0 +1,248 @@
+        for (SIZE_T i = 0; dst[i]; ++i)
+        {
+            if (dst[i] == L':' && lendir) // Don't allow absolute paths
+                dst[i] = L'_';
+            if (dst[i] == L'/') // Normalize
+                dst[i] = L'\\';
+        }
+    }
+    return path;
+}
+
+static HRESULT HResultFrom(const ERF &erf)
+{
+    switch (erf.fError ? erf.erfOper : FDIERROR_NONE)
+    {
+        case FDIERROR_NONE:
+            return erf.fError ? HRESULT_FROM_WIN32(erf.erfType) : S_OK;
+        case FDIERROR_CABINET_NOT_FOUND:
+            return HRESULT_FROM_WIN32(ERROR_FILE_NOT_FOUND);
+        case FDIERROR_ALLOC_FAIL:
+            return E_OUTOFMEMORY;
+        case FDIERROR_USER_ABORT:
+            return S_FALSE;
+        default:
+            return erf.erfType ? HRESULT_FROM_WIN32(erf.erfType) : E_FAIL;
+    }
+}
+
+FNFREE(CabMemFree)
+{
+    SHFree(pv);
+}
+
+FNALLOC(CabMemAlloc)
+{
+    return SHAlloc(cb);
+}
+
+FNCLOSE(CabClose)
+{
+    return CloseHandle((HANDLE)hf) ? 0 : -1;
+}
+
+static INT_PTR CabOpenEx(LPCWSTR path, UINT access, UINT share, UINT disp, UINT attr)
+{
+    return (INT_PTR)CreateFileW(path, access, share, NULL, disp, attr, NULL);
+}
+
+FNOPEN(CabOpen)
+{
+    UINT disp = (oflag & _O_CREAT) ? CREATE_ALWAYS : OPEN_EXISTING;
+    UINT access = GENERIC_READ;
+    if (oflag & _O_RDWR)
+        access = GENERIC_READ | GENERIC_WRITE;
+    else if (oflag & _O_WRONLY)
+        access = GENERIC_WRITE;
+    UNREFERENCED_PARAMETER(pmode);
+    WCHAR buf[MAX_PATH * 2];
+    MultiByteToWideChar(CP_UTF8, 0, pszFile, -1, buf, _countof(buf));
+    return CabOpenEx(buf, access, FILE_SHARE_READ, disp, FILE_ATTRIBUTE_NORMAL);
+}
+
+FNREAD(CabRead)
+{
+    DWORD dwBytesRead;
+    return ReadFile((HANDLE)hf, pv, cb, &dwBytesRead, NULL) ? dwBytesRead : -1;
+}
+
+FNWRITE(CabWrite)
+{
+    DWORD dwBytesWritten;
+    return WriteFile((HANDLE)hf, pv, cb, &dwBytesWritten, NULL) ? dwBytesWritten : -1;
+}
+
+FNSEEK(CabSeek)
+{
+    return SetFilePointer((HANDLE)hf, dist, NULL, seektype);
+}
+
+static HRESULT Init(HFDI &hfdi, ERF &erf)
+{
+    const int cpu = cpuUNKNOWN;
+    hfdi = FDICreate(CabMemAlloc, CabMemFree, CabOpen, CabRead, CabWrite, CabClose, CabSeek, cpu, &erf);
+    return hfdi ? S_OK : HResultFrom(erf);
+}
+
+FNFDINOTIFY(ExtractCabinetCallback)
+{
+    const UINT attrmask = FILE_ATTRIBUTE_HIDDEN | FILE_ATTRIBUTE_SYSTEM | FILE_ATTRIBUTE_ARCHIVE;
+    EXTRACTCABINETINTERNALDATA &ecd = *(EXTRACTCABINETINTERNALDATA*)pfdin->pv;
+    EXTRACTCALLBACKDATA noti;
+    HRESULT hr;
+    FILETIME ft;
+
+    noti.pfdin = pfdin;
+    switch (fdint)
+    {
+        case fdintCOPY_FILE:
+            hr = ecd.callback(ECM_FILE, noti, ecd.cookie);
+            if (hr == S_OK)
+            {
+                hr = E_OUTOFMEMORY;
+                LPWSTR path = BuildPath(ecd.destination, pfdin->psz1, pfdin->attribs);
+                if (path)
+                {
+                    // Callee is using SHPPFW_IGNOREFILENAME so we don't need to remove the name.
```

Comment:
It is always the case right now because the folders `ExtractFilesCallback` is the only one listening for this. I left the code there in case somebody wants to use it for something else. (This whole "callback API" is unique to extract.cpp and folder.cpp but could be used by other consumers of the CAB API if they want to copy this code)